Source/Core/Core/HW/DVD/DVDInterface.cpp

```diff
@@ -186,45 +190,57 @@ void DVDInterface::DTKStreamingCallback(DIInterruptType interrupt_type,
   // Actual games always set this to 48 KHz
   // but let's make sure to use GetAISSampleRateDivisor()
   // just in case it changes to 32 KHz
+  const auto sample_rate = ai.GetAISSampleRate();
   const u32 sample_rate_divisor = ai.GetAISSampleRateDivisor();
 
   // Determine which audio data to read next.
 
   // 3.5 ms of samples
-  const u32 maximum_samples =
-      ((Mixer::FIXED_SAMPLE_RATE_DIVIDEND / 2000) * 7) / sample_rate_divisor;
+  const u32 maximum_blocks = sample_rate == AudioInterface::SampleRate::AI32KHz ? 4 : 6;
   u64 read_offset = 0;
   u32 read_length = 0;
 
   if (interrupt_type == DIInterruptType::TCINT)
   {
+    constexpr u32 MAX_POSSIBLE_BLOCKS = 6;
+    constexpr u32 MAX_POSSIBLE_SAMPLES = MAX_POSSIBLE_BLOCKS * StreamADPCM::SAMPLES_PER_BLOCK;
+
     // Send audio to the mixer.
-    std::vector<s16> temp_pcm(m_pending_samples * 2, 0);
-    ProcessDTKSamples(&temp_pcm, audio_data);
+    std::array<s16, MAX_POSSIBLE_SAMPLES * 2> temp_pcm;
+    ASSERT(m_pending_blocks <= MAX_POSSIBLE_BLOCKS);
+    const u32 pending_blocks = std::min(m_pending_blocks, MAX_POSSIBLE_BLOCKS);
+    const size_t blocks_processed = ProcessDTKSamples(temp_pcm.data(), pending_blocks, audio_data);
+
+    // Fill unprocessed space up with zero samples.
```

Comment:
I think that's unnecessary outside of maybe debug mode, but it's so little (cached) memory anyway that it doesn't make a big performance difference. Your call, but I'd remove that, we already know exactly how much of that array was filled with proper initialized values.

If you really care, I'd just zero-initialize the array before ProcessDTKSamples -- that achieves mostly the same thing but doesn't introduce a data dependent loop.

Reply:
Yeah I *thought* this was unnecessary too at first, but when I tested it without the zero-fill I got some noise output in Ikaruga when the music started. See also #11654. I think you're right that it's probably better to just zero-initialize the whole array though.